File: src/main/java/de/zalando/zally/rules/PluralizeNamesForArraysRule.java
package de.zalando.zally.rules;

import de.zalando.zally.Violation;
import de.zalando.zally.ViolationType;
import io.swagger.models.Model;
import io.swagger.models.Swagger;
import io.swagger.models.properties.Property;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import static de.zalando.zally.utils.WordUtil.isPlural;

public class PluralizeNamesForArraysRule implements Rule {

    public static final String TITLE = "Array names should be pluralized";
    public static final String DESC_PATTERN = "Array property name '%s' is not pluralized (but we are not sure)";
    public static final String RULE_URL = "http://zalando.github.io/restful-api-guidelines/json-guidelines/" +
            "JsonGuidelines.html#should-array-names-should-be-pluralized";

    @Override
    public List<Violation> validate(Swagger swagger) {
        List<Violation> res = new ArrayList<>();
        if (swagger !=null && swagger.getDefinitions() != null) {
            for (Map.Entry<String, Model> entry : swagger.getDefinitions().entrySet()) {
                String definition = entry.getKey();
                Model model = entry.getValue();
                if (model != null && model.getProperties() != null) {
                    for (Map.Entry<String, Property> e : model.getProperties().entrySet()) {
                        String name = e.getKey();
                        if ("array".equals(e.getValue().getType()) && !isPlural(name)) {
                            res.add(new Violation(
                                    TITLE,
                                    String.format(DESC_PATTERN, name),
                                    ViolationType.SHOULD, RULE_URL, definition + "." + name));
                        }
                    }
                }
            }
        }
        return res;
    }
}

File: server/src/main/java/de/zalando/zally/rules/PluralizeNamesForArraysRule.kt
package de.zalando.zally.rules

import de.zalando.zally.Violation
import de.zalando.zally.ViolationType
import de.zalando.zally.utils.WordUtil.isPlural
import io.swagger.models.Swagger
import org.springframework.stereotype.Component

@Component
class PluralizeNamesForArraysRule : AbstractRule() {
    override val title = "Array names should be pluralized"
    override val url = "http://zalando.github.io/restful-api-guidelines/json-guidelines/JsonGuidelines.html" +
            "#should-array-names-should-be-pluralized"
    override val violationType = ViolationType.SHOULD
    override val code = "S007"

    override fun validate(swagger: Swagger): Violation? {
        val res = swagger.definitions.orEmpty().entries.map { def ->
            val badProps = def.value?.properties.orEmpty().entries.filter { "array" == it.value?.type && !isPlural(it.key) }
            if (badProps.isNotEmpty()) {
                val propsDesc = badProps.map { "'${it.key}'" }.joinToString(",")
                "Definition ${def.key}: $propsDesc" to "#/definitions/${def.key}"
            } else null
        }.filterNotNull()

        return if (res.isNotEmpty()) {
            val (desc, paths) = res.unzip()
            Violation(this, title, desc.joinToString("\n"), violationType, url, paths)
        } else null
    }
}
